libc/stdlib_alloc.c

```diff
@@ -0,0 +1,277 @@
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(float), malloc_usable_size(ptr1));
+#endif
+	void *ptr2;
+	ptr2 = malloc(sizeof(char));
+	TEST_ASSERT_NOT_NULL(ptr2);
+	TEST_ASSERT_TRUE(ptr2 != 0);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(char), malloc_usable_size(ptr2));
+#endif
+	free(ptr0);
+	free(ptr1);
+	free(ptr2);
+}
+
+
+TEST(stdlib_malloc, malloc_zero)
+{
+	void *ptr;
+	ptr = malloc(0);
+	TEST_ASSERT_TRUE(ptr == NULL || ptr != 0);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_basic)
+{
+	int *ptr = calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_TRUE(ptr != 0);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(BLOCK_SIZE, malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_zero)
+{
+	int *ptr0 = calloc(0, sizeof(int));
+	TEST_ASSERT_TRUE(ptr0 == NULL || ptr0 != 0);
+	int *ptr1 = calloc(BLOCK_SIZE, 0);
+	TEST_ASSERT_TRUE(ptr1 == NULL || ptr1 != 0);
+	free(ptr0);
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, calloc_one)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_large)
+{
+	int *ptr = calloc(BLOCK_SIZE * LARGE_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_TRUE(ptr != 0);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE * LARGE_SIZE);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, realloc_null)
+{
+	int *ptr = realloc(NULL, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, realloc_zero_size)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	int *ptr1 = realloc(ptr, 0);
+	TEST_ASSERT_NULL(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize)
+{
+	int NUMBERS[BLOCK_SIZE];
+	int *ptr = (int *)calloc(BLOCK_SIZE, sizeof(int));
+	for (int i = 0; i < BLOCK_SIZE; i++) {
+		ptr[i] = i;
+		NUMBERS[i] = i;
+	}
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr, BLOCK_SIZE);
+
+	int *ptr1 = (int *)realloc(ptr, BLOCK_SIZE * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr1, BLOCK_SIZE);
+
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize_small)
+{
+	int *ptr = (int *)calloc(1, 2 * sizeof(int));
+	ptr[0] = 1;
+
+	int *ptr1 = (int *)realloc(ptr, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT(1, ptr1[0]);
+
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize_large)
+{
+	int *ptr = (int *)calloc(BLOCK_SIZE, 2 * sizeof(int));
+
+	int *ptr1 = (int *)realloc(ptr, LARGE_SIZE * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_multiple)
+{
+	int *ptr = calloc(1, BLOCK_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	int *ptr1 = realloc(ptr, BLOCK_SIZE);
+	TEST_ASSERT_NOT_NULL(ptr1);
+	ptr = NULL;
+	TEST_ASSERT_NULL(ptr);
+	TEST_ASSERT_TRUE(ptr != ptr1);
```

Comment:
There is no need for these checks. We made them already - `ptr1` is not NULL and we set `ptr` to NULL.
Applies also below